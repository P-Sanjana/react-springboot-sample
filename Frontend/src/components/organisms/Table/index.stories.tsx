import { Meta, Story } from "@storybook/react";
import React, { ComponentProps } from "react";
import TradeTable from "./index";
import { selectedCrypto } from "../AmountDetails/index.stories";

export default {
  title: "organisms/Table",
  component: TradeTable,
} as Meta;

const Template: Story<ComponentProps<typeof TradeTable>> = (args) => (
  <TradeTable {...args} />
);

export const Default = Template.bind({});
Default.args = {
  setWatch: (di) => console.log(`Set watch ${di}`),
  data: [selectedCrypto],
  onSort: (field, dir) => console.log(`Sort called for ${field} ${dir}`),
};
