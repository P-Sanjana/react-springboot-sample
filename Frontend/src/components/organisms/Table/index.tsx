import {
  Box,
  ButtonBase,
  IconButton,
  makeStyles,
  styled,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from "@material-ui/core";
import { Star, StarBorderOutlined } from "@material-ui/icons";
import React, { useState } from "react";
import upDownArrow from "../../../assets/icons/code-line.svg";
import { CurrencySummary } from "../../../types/currency.type";
import { Currency } from "../../../types/responseTypes";
import localeUtils from "../../../utils/locale-utils";
import {
  FieldPicker,
  SortDirection,
  SortEventHandler,
  tradeTableHeaders,
} from "../../../utils/table-utils";

const useStyles = makeStyles((theme) => ({
  tableRow: {
    boxSizing: "border-box",
    borderRadius: theme.spacing(1),
    backgroundColor: theme.palette.background.paper,
    "&:hover": {
      boxShadow: `0px 2px 4px ${theme.palette.grey[300]}`,
    },
  },
  tableStartCell: {
    borderLeft: `1px solid ${theme.palette.grey[100]}`,
  },
  tableEndCell: {
    borderRight: `1px solid ${theme.palette.grey[100]}`,
  },
  table: {
    borderCollapse: "separate",
    borderSpacing: theme.spacing(0, 3),
  },
  changeStylePositive: {
    color: theme.palette.success[500],
  },
  changeStyleNegative: {
    color: theme.palette.error[500],
  },
  currency: {
    display: "flex",
    flexDirection: "column",
    paddingLeft: theme.spacing(1),
  },
  currencyCode: {
    color: `${theme.palette.text.secondary}`,
  },
  currencyImage: {
    width: "42px",
    height: "42px",
    borderRadius: "100px",
  },
  nameCellContent: {
    marginLeft: theme.spacing(4),
  },
}));

const TradeTableHeaderCell = styled(TableCell)({
  border: "none",
});

const TradeTableDataCell = styled(TableCell)(({ theme }) => ({
  borderTop: `1px solid ${theme.palette.grey[100]}`,
  borderBottom: `1px solid ${theme.palette.grey[100]}`,
}));

const HeaderTableCellTypography = styled(Typography)(({ theme }) => ({
  color: theme.palette.grey["500"],
}));

export type SetWatchEventHandler = (dataIndex: number, watch: boolean) => void;

export type TableProps = {
  data: Array<Currency>;
  setWatch: SetWatchEventHandler;
  onSort?: SortEventHandler<Currency>;
};

const TradeTable: React.FC<TableProps> = ({ data, setWatch, onSort }) => {
  const toggleWatch = (dataIndex: number) => {
    setWatch(dataIndex, !data[dataIndex].watch);
  };

  const [sortDirections, setSortDirections] = useState(
    new Map<string, SortDirection>(
      tradeTableHeaders
        .filter(({ sort }) => sort)
        .map(({ label }): [string, SortDirection] => [label, "none"]),
    ),
  );

  const toggleSort = (label: string, fieldPicker: FieldPicker<Currency>) => {
    const newSortDirections = new Map(sortDirections);
    const currDir = newSortDirections.get(label);
    let newDir: SortDirection = "none";
    if (currDir == "none") {
      newDir = "asc";
    } else if (currDir == "asc") {
      newDir = "desc";
    }
    newSortDirections.set(label, newDir);
    setSortDirections(newSortDirections);
    if (onSort) {
      onSort(fieldPicker, newDir);
    }
  };

  const classes = useStyles();

  const tableBody = data.map(
    (
      {
        logoUrl,
        name,
        price,
        changeRate,
        marketCap,
        watch,
        code,
        currencyType,
      },
      index,
    ) => {
      const changeStyleCN =
        changeRate >= 0
          ? classes.changeStylePositive
          : classes.changeStyleNegative;
      const changeSign = changeRate > 0 ? "+" : "";
      if (currencyType === "CRYPTO")
        return (
          <TableRow className={classes.tableRow} key={index}>
            <TradeTableDataCell className={classes.tableStartCell}>
              <Box className={classes.nameCellContent} display="flex">
                <img
                  src={logoUrl}
                  className={classes.currencyImage}
                  alt="currency image"
                />

                <Box className={classes.currency}>
                  <Typography variant="body1">{name}</Typography>

                  <Typography
                    variant="overline"
                    className={classes.currencyCode}
                  >
                    {code}
                  </Typography>
                </Box>
              </Box>
            </TradeTableDataCell>
            <TradeTableDataCell>
              <Typography variant="body2" data-testid="price">
                {localeUtils.formatCurr(price)}
              </Typography>
            </TradeTableDataCell>

            <TradeTableDataCell>
              <Typography
                className={changeStyleCN}
                variant="body2"
                data-testid="change"
              >
                {changeSign}
                {changeRate}%
              </Typography>
            </TradeTableDataCell>

            <TradeTableDataCell>
              <Typography variant="body2" data-testid="marketCap">
                {localeUtils.abbreviateCurrency(marketCap)}
              </Typography>
            </TradeTableDataCell>

            <TradeTableDataCell className={classes.tableEndCell}>
              <IconButton
                onClick={() => toggleWatch(index)}
                data-testid="watch-toggle"
              >
                {watch ? (
                  <Star data-testid="watch" color="primary" />
                ) : (
                  <StarBorderOutlined data-testid="watch" />
                )}
              </IconButton>
            </TradeTableDataCell>
          </TableRow>
        );
    },
  );

  const tableHeader = tradeTableHeaders.map(
    ({ label, sort, sortFieldPicker }, i) => {
      let cellContent;
      if (sort && sortFieldPicker) {
        cellContent = (
          <ButtonBase
            onClick={() => toggleSort(label, sortFieldPicker)}
            data-testid="sort-toggle"
          >
            <Box display="flex" flexDirection="row" alignItems="center">
              <HeaderTableCellTypography variant="caption">
                {label}
              </HeaderTableCellTypography>
              <img
                src={upDownArrow}
                alt=""
                style={{ transform: "rotate(90deg)" }}
                height="15px"
                width="15px"
              />
            </Box>
          </ButtonBase>
        );
      } else {
        cellContent = (
          <HeaderTableCellTypography
            variant="caption"
            className={i == 0 ? classes.nameCellContent : undefined}
          >
            {label}
          </HeaderTableCellTypography>
        );
      }
      return <TradeTableHeaderCell key={i}>{cellContent}</TradeTableHeaderCell>;
    },
  );

  return (
    <TableContainer>
      <Table className={classes.table}>
        <TableHead>
          <TableRow>{tableHeader}</TableRow>
        </TableHead>
        <TableBody>{tableBody}</TableBody>
      </Table>
    </TableContainer>
  );
};

export default TradeTable;
