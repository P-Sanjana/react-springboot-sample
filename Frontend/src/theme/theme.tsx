import { Color } from "@material-ui/core";
import { createTheme } from "@material-ui/core/styles";
import GraphikMedium from "../assets/fonts/GraphikMedium.otf";
import GraphikRegular from "../assets/fonts/GraphikRegular.otf";
import GraphikSemiBold from "../assets/fonts/GraphikSemibold.otf";

type ColorPartial = Partial<Color>;

declare module "@material-ui/core/styles/createPalette" {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface PaletteColor extends ColorPartial {}
}

declare module "@material-ui/core/styles/createTypography" {
  interface Typography {
    inhherit: React.CSSProperties;
    caption2: React.CSSProperties;
  }

  interface TypographyOptions {
    inherit?: React.CSSProperties;
    caption2?: React.CSSProperties;
  }
}

declare module "@material-ui/core/Typography/Typography" {
  interface TypographyPropsVariantOverrides {
    caption2: true;
  }
}

const fontFaceGraphikRegular = {
  fontFamily: "GraphikRegular",
  src: `url(${GraphikRegular}) format("opentype")`,
};

const fontFaceGraphikSemiBold = {
  fontFamily: "GraphikSemiBold",
  src: `url(${GraphikSemiBold}) format("opentype")`,
};

const fontFaceGraphikMedium = {
  fontFamily: "GraphikMedium",
  src: `url(${GraphikMedium}) format("opentype")`,
};

const theme = createTheme({
  spacing: [0, 4, 8, 12, 16, 20, 24, 32],
  palette: {
    primary: {
      main: "#0052FF",
      "100": "#FAFCFF",
      "300": "#CCE3FF",
      "500": "#0052FF",
      "700": "#002EB7",
      "900": "#00177A",
    },
    success: {
      main: "#20B03F",
      "500": "#20B03F",
      "100": "#E9F7EC",
    },
    warning: {
      main: "#FFA74F",
      "500": "#FFA74F",
      "100": "#FFF6ED",
    },
    error: {
      main: "#B71A33",
      "500": "#B71A33",
      "100": "#F3E6EB",
    },
    grey: {
      "50": "#F2F2F7",
      "100": "#E8E8F7",
      "300": "#B4B4CF",
      "500": "#4B4B60",
      "700": "#252545",
      "900": "#0E0E2E",
    },
    text: {
      primary: "#343446",
      secondary: "#7D7D89",
      hint: "#7D7D89",
      disabled: "#B2B2B9",
    },
    secondary: {
      main: "#FFA74F",
      contrastText: "#FFFFFF",
    },
  },
  overrides: {
    MuiCssBaseline: {
      "@global": {
        "@font-face": [
          fontFaceGraphikRegular,
          fontFaceGraphikSemiBold,
          fontFaceGraphikMedium,
        ],
      },
    },
  },
  typography: {
    overline: {
      gammaOverlineFontFamily: "Comic Sans Ms ",
      fontStyle: "normal",
      fontWeight: "normal",
      fontSize: "12px",
      lineHeight: "14px",
      identicalToBoxHeightOr117LetterSpacing: "0.005em",
      betaTextHighEmphasisColor: "#343446",
    },
    h4: {
      fontFamily: "GraphikMedium",
      fontSize: "40px",
      fontWeight: 500,
      letterSpacing: "-1%",
      lineHeight: "54px",
    },
    h6: {
      fontFamily: "GraphikRegular",
      fontStyle: "normal",
      fontWeight: 400,
      fontSize: "24px",
      lineHeight: "34px",
    },
    caption: {
      fontFamily: "GraphikMedium",
      fontStyle: "normal",
      fontWeight: 500,
      fontSize: "14px",
      lineHeight: "16px",
      letterSpacing: "0.01em",
    },
    inherit: {
      fontFamily: "GraphikRegular",
      fontStyle: "normal",
      fontWeight: "normal",
      fontSize: "14px",
      lineHeight: "16px",
      letterSpacing: "0.01em",
    },
    button: {
      fontFamily: "GraphikSemiBold",
      fontStyle: "normal",
      fontWeight: 500,
      fontSize: "14px",
      lineHeight: "42px",
      letterSpacing: "0.01em",
    },
    body2: {
      fontFamily: "GraphikRegular",
      fontStyle: "normal",
      fontWeight: "normal",
      fontSize: "16px",
      lineHeight: "22px",
      letterSpacing: "0.01em",
    },
    subtitle2: {
      fontFamily: "GraphikRegular",
      fontStyle: "normal",
      fontWeight: "normal",
      fontSize: "20px",
      lineHeight: "28px",
      letterSpacing: "0.005em",
    },
    subtitle1: {
      fontFamily: "GraphikSemiBold",
      fontStyle: "normal",
      fontWeight: 500,
      fontSize: "20px",
      lineHeight: "28px",
      letterSpacing: "0.005em",
    },
    body1: {
      fontFamily: "GraphikSemiBold",
      fontStyle: "normal",
      fontWeight: 500,
      fontSize: "16px",
      lineHeight: "22px",
      letterSpacing: "0.01em",
    },
    caption2: {
      fontFamily: "GraphikRegular",
      fontStyle: "normal",
      fontWeight: 400,
      fontSize: "14px",
      lineHeight: "16px",
      letterSpacing: "0.01em",
    },
  },
});

export default theme;
