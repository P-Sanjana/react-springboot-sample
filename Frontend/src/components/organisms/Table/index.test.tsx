import { ThemeProvider } from "@material-ui/core";
import { render } from "@testing-library/react";
import React from "react";
import theme from "../../../theme/theme";
import Table, { TableProps } from "./index";

describe("Table Component", () => {
  const commonTableProps: TableProps = {
    setWatch: (di) => console.log(`Set watch ${di}`),
    data: [
      {
        id: 1,
        name: "Bitcoin",
        code: "BTC",
        color: "#F7931A",
        logoUrl:
          "https://s3-alpha-sig.figma.com/img/3081/886d/bb95f08981a385cf742cbde18bac74e1?Expires=1635724800&Signature=HW8nL9s0E8pc8bHC5FFlxSzUHlSavGAzYHsKxFS7KpTsVaMQ~V4HjAoKP5HggTq18zXZkiKCtt9wCxDJ9J0BvHPjp90jagsn8n3j5rQTsqtuAblTpbNaPEBcJ5h-fr5zhJ2eFe1H64iOF7DdwMaaEeMCTcgJUOJhHzVVg2YkbEsybC0j~WhqfHX9d-H4TojDJYKt-lF4ztnQ17ZBEBOJem-h1y6B3fSY2K2B83YVGWgwqHKdww7RqUVQrKljDxJ7nGJnpUiWmn7R8mjqULxr7160-YGDJV-cGlOed7nMSAM1F9ARfZWOWweT0T6aP3FOc3E3zjqh9ycAP6mCiqUUMg__&Key-Pair-Id=APKAINTVSUGEWH5XD5UA",
        watch: true,
        price: 61048.7,
        priceSymbol: "$",
        changeRate: 1.06,
        marketCapStr: "$1.15T",
        marketCap: 1149846087372,
        vol24h: "$24.7B",
        circulatingSupply: "18M BTC",
      },
    ],
  };
  it("should have semantic2 color when change is positive", () => {
    const tableProps = { ...commonTableProps };
    tableProps.data[0].changeRate = 2.9;
    const wrapper = render(
      <ThemeProvider theme={theme}>
        <Table {...tableProps} />
        );
      </ThemeProvider>,
    );
    const comp = wrapper.getByTestId("change");
    expect(comp).toHaveStyle({ color: theme.palette.success.main });
  });
  it("should have semantic6 color when change is negative", () => {
    const tableProps = { ...commonTableProps };
    tableProps.data[0].changeRate = -2.9;
    const wrapper = render(
      <ThemeProvider theme={theme}>
        <Table {...tableProps} />
        );
      </ThemeProvider>,
    );
    const comp = wrapper.getByTestId("change");
    expect(comp).toHaveStyle({ color: theme.palette.error.main });
  });
  it("should match snapshot", () => {
    const wrapper = render(
      <ThemeProvider theme={theme}>
        <Table {...commonTableProps} />
        );
      </ThemeProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
