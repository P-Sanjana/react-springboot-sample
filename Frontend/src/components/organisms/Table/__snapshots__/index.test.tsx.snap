// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTableContainer-root"
      >
        <table
          class="MuiTable-root makeStyles-table-56"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
                scope="col"
              >
                <span
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-65 makeStyles-nameCellContent-62 MuiTypography-caption"
                >
                  Name
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
                scope="col"
              >
                <span
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-66 MuiTypography-caption"
                >
                  Price
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
                scope="col"
              >
                <span
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-67 MuiTypography-caption"
                >
                  Change
                </span>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
                scope="col"
              >
                <button
                  class="MuiButtonBase-root"
                  data-testid="sort-toggle"
                  tabindex="0"
                  type="button"
                >
                  <div
                    class="MuiBox-root MuiBox-root-68"
                  >
                    <span
                      class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-69 MuiTypography-caption"
                    >
                      MarketCap
                    </span>
                    <img
                      alt=""
                      height="15px"
                      src="code-line.svg"
                      style="transform: rotate(90deg);"
                      width="15px"
                    />
                  </div>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
                scope="col"
              >
                <span
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-70 MuiTypography-caption"
                >
                  Watch
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root makeStyles-tableRow-53"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-72 makeStyles-tableStartCell-54"
              >
                <div
                  class="MuiBox-root MuiBox-root-73 makeStyles-nameCellContent-62"
                >
                  <img
                    alt="currency image"
                    class="makeStyles-currencyImage-61"
                    src="https://s3-alpha-sig.figma.com/img/3081/886d/bb95f08981a385cf742cbde18bac74e1?Expires=1635724800&Signature=HW8nL9s0E8pc8bHC5FFlxSzUHlSavGAzYHsKxFS7KpTsVaMQ~V4HjAoKP5HggTq18zXZkiKCtt9wCxDJ9J0BvHPjp90jagsn8n3j5rQTsqtuAblTpbNaPEBcJ5h-fr5zhJ2eFe1H64iOF7DdwMaaEeMCTcgJUOJhHzVVg2YkbEsybC0j~WhqfHX9d-H4TojDJYKt-lF4ztnQ17ZBEBOJem-h1y6B3fSY2K2B83YVGWgwqHKdww7RqUVQrKljDxJ7nGJnpUiWmn7R8mjqULxr7160-YGDJV-cGlOed7nMSAM1F9ARfZWOWweT0T6aP3FOc3E3zjqh9ycAP6mCiqUUMg__&Key-Pair-Id=APKAINTVSUGEWH5XD5UA"
                  />
                  <div
                    class="MuiBox-root MuiBox-root-74 makeStyles-currency-59"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                    >
                      Bitcoin
                    </p>
                    <span
                      class="MuiTypography-root makeStyles-currencyCode-60 MuiTypography-caption"
                    >
                      BTC
                    </span>
                  </div>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-75"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2"
                  data-testid="price"
                >
                  $
                  61,048.7
                </p>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-76"
              >
                <p
                  class="MuiTypography-root makeStyles-changeStyleNegative-58 MuiTypography-body2"
                  data-testid="change"
                >
                  
                  -2.9
                  %
                </p>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-77"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2"
                  data-testid="marketCap"
                >
                  $1.15T
                </p>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-78 makeStyles-tableEndCell-55"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  data-testid="watch-toggle"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                      data-testid="watch"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      );
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTableContainer-root"
    >
      <table
        class="MuiTable-root makeStyles-table-56"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
              scope="col"
            >
              <span
                class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-65 makeStyles-nameCellContent-62 MuiTypography-caption"
              >
                Name
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
              scope="col"
            >
              <span
                class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-66 MuiTypography-caption"
              >
                Price
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
              scope="col"
            >
              <span
                class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-67 MuiTypography-caption"
              >
                Change
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
              scope="col"
            >
              <button
                class="MuiButtonBase-root"
                data-testid="sort-toggle"
                tabindex="0"
                type="button"
              >
                <div
                  class="MuiBox-root MuiBox-root-68"
                >
                  <span
                    class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-69 MuiTypography-caption"
                  >
                    MarketCap
                  </span>
                  <img
                    alt=""
                    height="15px"
                    src="code-line.svg"
                    style="transform: rotate(90deg);"
                    width="15px"
                  />
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-root-63"
              scope="col"
            >
              <span
                class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-64 WithStyles(ForwardRef(Typography))-root-70 MuiTypography-caption"
              >
                Watch
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root makeStyles-tableRow-53"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-72 makeStyles-tableStartCell-54"
            >
              <div
                class="MuiBox-root MuiBox-root-73 makeStyles-nameCellContent-62"
              >
                <img
                  alt="currency image"
                  class="makeStyles-currencyImage-61"
                  src="https://s3-alpha-sig.figma.com/img/3081/886d/bb95f08981a385cf742cbde18bac74e1?Expires=1635724800&Signature=HW8nL9s0E8pc8bHC5FFlxSzUHlSavGAzYHsKxFS7KpTsVaMQ~V4HjAoKP5HggTq18zXZkiKCtt9wCxDJ9J0BvHPjp90jagsn8n3j5rQTsqtuAblTpbNaPEBcJ5h-fr5zhJ2eFe1H64iOF7DdwMaaEeMCTcgJUOJhHzVVg2YkbEsybC0j~WhqfHX9d-H4TojDJYKt-lF4ztnQ17ZBEBOJem-h1y6B3fSY2K2B83YVGWgwqHKdww7RqUVQrKljDxJ7nGJnpUiWmn7R8mjqULxr7160-YGDJV-cGlOed7nMSAM1F9ARfZWOWweT0T6aP3FOc3E3zjqh9ycAP6mCiqUUMg__&Key-Pair-Id=APKAINTVSUGEWH5XD5UA"
                />
                <div
                  class="MuiBox-root MuiBox-root-74 makeStyles-currency-59"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1"
                  >
                    Bitcoin
                  </p>
                  <span
                    class="MuiTypography-root makeStyles-currencyCode-60 MuiTypography-caption"
                  >
                    BTC
                  </span>
                </div>
              </div>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-75"
            >
              <p
                class="MuiTypography-root MuiTypography-body2"
                data-testid="price"
              >
                $
                61,048.7
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-76"
            >
              <p
                class="MuiTypography-root makeStyles-changeStyleNegative-58 MuiTypography-body2"
                data-testid="change"
              >
                
                -2.9
                %
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-77"
            >
              <p
                class="MuiTypography-root MuiTypography-body2"
                data-testid="marketCap"
              >
                $1.15T
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body WithStyles(ForwardRef(TableCell))-root-71 WithStyles(ForwardRef(TableCell))-root-78 makeStyles-tableEndCell-55"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="watch-toggle"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    data-testid="watch"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    );
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
